### PR TITLE
Make Preprocessor aware of the LF version to process upgrade

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
instead of using a flag.

### CODE REVIEW

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 1]
```diff
@@ -84,91 +79,189 @@ private[engine] final class Preprocessor(
...
+
+  private[this] def collectPkgInDisclosure(
+      contracts: ImmArray[command.DisclosedContract]
+  ): Set[language.Reference] =
+    contracts.iterator
+      .map(contract => language.Reference.Template(TypeConRef.fromIdentifier(contract.templateId)))
+      .toSet
 
-  private[this] def pullPackage(
+  private[this] def pullPackages(
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:215]: Took me a while to realize that compiledPackages was mutable and this was having a side effect. Maybe add a small comment along the lines of "populates compiledPackages".

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 2]
```diff
@@ -337,13 +255,6 @@ private[lf] final class CommandPreprocessor(
           contractKey,
           choiceId,
           argument,
-          strictArgument = true,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:340]: This made me realize that, if I understand correctly, we're now also more lenient for arguments during replay in certain cases. Is that ok?

File: sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala [hunk 3]
```diff
@@ -21,360 +20,440 @@ import com.daml.lf.speedy.Compiler
...
+
+    val testCases = Table[Ast.Type, Value, List[Value], speedy.SValue](
+      ("type", "normalized value", "non normalized value", "svalue"),
       (
-        t"Mod:Tuple Int64 Text",
-        ValueRecord("", ImmArray("x" -> ValueInt64(33), "y" -> ValueText("a"))),
-        SRecord("Mod:Tuple", ImmArray("x", "y"), ArrayList(SInt64(33), SText("a"))),
+        t"Mod:Record Int64 Text Party Boolean Unit",
+        ValueRecord("", ImmArray("" -> aInt, "" -> none, "" -> someParty)),
+        List(
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala:141]: As discussed over zoom, add a case where the package ID doesn't resolve to the same package name and show that it is ignored as intended. Same below for variants and enums.

File: sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala [hunk 4]
```diff
@@ -54,68 +54,83 @@ object LookupError {
 
 sealed abstract class Reference extends Product with Serializable {
```
Comment (paulbrauner-da) [inline sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala:55]: Now that this isn't only used for error reporting but also for looking up packages in the preprocessor, should we move it out of LookupError.scala?

File: sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala [hunk 5]
```diff
@@ -287,15 +284,6 @@ private[script] object RunnerMainConfigIntermediate {
         s"Uploads the dar before running. Only available over GRPC. Default behaviour is to upload with --all, not with --script-name."
       )
 
-    opt[Unit]("enable-contract-upgrading")
```
Comment (paulbrauner-da) [inline sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala:290]: We may have to update the doc.

File: sdk/daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala [hunk 6]
```diff
@@ -56,7 +56,7 @@ class UpgradesIT extends AsyncWordSpec with AbstractScriptTest with Inside with
   // Maybe provide our own tracer that doesn't tag, it makes the logs very long
   "Multi-participant Daml Script Upgrades" should {
     testCases.foreach { testCase =>
```
Comment (paulbrauner-da) [inline sdk/daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala:58]: Was that an omission that you discovered while touching this file or does it break if you don't filter? (And if it does, why?)

File: sdk/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala [hunk 7]
```diff
@@ -53,15 +53,15 @@ final class Converter(
 
   private[this] val valueTranslator = new preprocessing.ValueTranslator(
     compiledPackages.pkgInterface,
-    requireV1ContractIdSuffix = false,
+    checkV1ContractIdSuffixes = true,
```
Comment (paulbrauner-da) [inline sdk/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:56]: Why the change here?

**Review by paulbrauner-da — APPROVED** [review]
Very nice!

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 8]
```diff
@@ -47,31 +48,25 @@ private[engine] final class Preprocessor(
...
       pkgInterface = pkgInterface,
-      requireV1ContractIdSuffix = requireV1ContractIdSuffix,
+      checkV1ContractIdSuffix = requireV1ContractIdSuffix,
     )
 
   val transactionPreprocessor = new TransactionPreprocessor(commandPreprocessor)
 
   @tailrec
-  private[this] def collectNewPackagesFromTypes(
+  private[this] def collectPackagesRefFromTypes(
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:57]: I just realize that if I ignore the packageId in the value, I just do not need any of the change in Preprocessor.
I will revert that.

File: sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala:290]: where it is ?

File: sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala:55]: I should revert that as well.

File: sdk/daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply sdk/daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala]: I forgot to revert it.

File: sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala (hunk 5 above)
Comment (paulbrauner-da) [reply sdk/daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMainConfig.scala:290]: [here](https://github.com/digital-asset/docs.daml.com/blob/d46ff4ae96dedf8c3f91aaaf76ecb3c4adc72be5/docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst?plain=1#L605) and below in the same document.

File: sdk/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply sdk/triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala]: not very important. reverted

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:340]: In 2.x, i do not think it is a big risk taht the predecessor is bit lenient.
In fact, it is already lenient from some other aspects (e.g. it does not require map to me normalize, ...)
In 3.x, I have done a dedicate [PR](https://github.com/digital-asset/daml/pull/20036) to enforce normalization.
